types/src/v2/params.rs

```diff
@@ -89,40 +89,39 @@ impl Serialize for TwoPointZero {
 /// The data containing the params is a `Cow<&str>` and can either be a borrowed `&str` of JSON from an incoming
 /// [`super::request::JsonRpcRequest`] (which in turn borrows it from the input buffer that is shared between requests);
 /// or, it can be an owned `String`.
+///
+/// Regards empty `JSON array` and `JSON object` as no parameters provided.
 #[derive(Clone, Debug)]
-pub struct RpcParams<'a>(Option<Cow<'a, str>>);
+pub struct RpcParams<'a>(Cow<'a, str>);
 
 impl<'a> RpcParams<'a> {
 	/// Create params
 	pub fn new(raw: Option<&'a str>) -> Self {
-		Self(raw.map(Into::into))
+		Self(match raw.unwrap_or("").trim() {
+			"{}" | "[]" => "".into(),
+			other => other.into(),
+		})
 	}
 
 	/// Returns true if the contained JSON is an object
 	pub fn is_object(&self) -> bool {
-		let json: &str = match self.0 {
-			Some(ref cow) => cow,
-			None => return false,
-		};
-
-		json.trim_start().starts_with('{')
+		self.0.starts_with('{')
```

Comment:
trimmed in the `new`

Reply:
With the string always being pre-trimmed, you can also move the trim in `next_inner` from line 156 to 172 (`trim_start` next subslice), which will skip one trim for the first param.